# Collecting the data set for NHL goalies from 1950-2021

Goalies are one of the most important, if not *the* most important position on a hockey team.  Even if the team is playing well in front of a goalie, the goalie can let in several soft goals and lose the game.  Conversely, if the goalie plays very well, that can give the team confidence and swagger, which can spur them on to playing well down the ice.  That being the case, it is crucial for a team to have a good goaltender if they want to win the stanley cup.  The purpose of this data science project is to collect a large amount of data for goaltenders and discover trends within the data.  Ideally, we would like to gain some insights into: who tend to be the best goaltenders; if the goaltending statistics that make a good goalie have changed over time; whether age has a large impact on play; other considerations.  We then plan on using producing a predictive model with the intent on being able to predict whether a goalie will be good and how many years he may be good for, in order to inform decisions such as contract term or whether to sign a goalie at all.

First thing is first, we will import the main data in the form of 41 excel spreadsheets downloaded from <a href="http://www.nhl.com/stats/goalies"> the NHL.com website stats tab </a>

In [1]:
import pandas as pd

The files we have are in excel spread sheet form.  Indeed, in 41 of them.  We will loop and read the data into a list of databases, then consolidate those databases into one.  They are all in the Downloads folder on my computer and have the end filename format of "Summary(x).xlsx"

In [11]:
baseFilePath = "/Users/codypewarchuk/Downloads/Summary "
listOfDataFrames = []
for i in range(41):
    df = pd.read_excel(baseFilePath + '(' + str(i) + ').xlsx')
    listOfDataFrames.append(df)
print(len(listOfDataFrames))

41


Now let's take those 41 dataframes and combine them into one to facilitate our analysis.

In [12]:
goalieDf = pd.concat(listOfDataFrames, ignore_index = True)

Let's check out our new dataframe's first 20 elements.

In [13]:
goalieDf.head(20)

Player    Season Team S/C  GP  GS   W   L   T  OT  ...    Svs  \
0        Braden Holtby  20152016  WSH   L  66  66  48   9  --   7  ...  1,661   
1       Martin Brodeur  20062007  NJD   L  78  78  48  23  --   7  ...  2,011   
2        Bernie Parent  19731974  PHI   L  73  73  47  13  12  --  ...  1,870   
3       Roberto Luongo  20062007  VAN   L  76  75  47  22  --   6  ...  1,998   
4       Evgeni Nabokov  20072008  SJS   L  77  77  46  21  --   8  ...  1,639   
5       Martin Brodeur  20092010  NJD   L  77  76  45  25  --   6  ...  1,836   
6     Miikka Kiprusoff  20082009  CGY   L  76  76  45  24  --   5  ...  1,946   
7          Carey Price  20142015  MTL   L  66  66  44  16  --   6  ...  1,823   
8    Connor Hellebuyck  20172018  WPG   L  67  64  44  11  --   9  ...  1,892   
9       Evgeni Nabokov  20092010  SJS   L  71  71  44  16  --  10  ...  1,998   
10      Martin Brodeur  20072008  NJD   L  77  77  44  27  --   6  ...  1,921   
11  Andrei Vasilevskiy  20172018  TBL   L  65  64  44  17  --   3  ...  1,908   
12       Bernie Parent  19741975  PHI   L  68  68  44  14   9  --  ...  1,537   
13       Terry Sawchuk  19501951  DET   L  70  70  44  13  13  --  ...     --   
14       Terry Sawchuk  19511952  DET   L  70  70  44  14  12  --  ...     --   
15         Pekka Rinne  20112012  NSH   L  73  72  43  18  --   8  ...  1,987   
16      Martin Brodeur  19971998  NJD   L  70  69  43  17   8  --  ...  1,439   
17      Martin Brodeur  20052006  NJD   L  73  73  43  23  --   7  ...  1,918   
18      Martin Brodeur  19992000  NJD   L  72  72  43  20   8  --  ...  1,636   
19          Ed Belfour  19901991  CHI   L  74  70  43  19   7  --  ...  1,713   

     GA    Sv%   GAA      TOI  SO  G  A  P  PIM  
0   141  0.922  2.20  3841:11   3  0  1  1    6  
1   171  0.922  2.18  4696:33  12  0  1  1   12  
2   136  0.932  1.89  4307:25  12  0  3  3   24  
3   171  0.921  2.28  4490:16   5  0  2  2   10  
4   163   0.91  2.14  4560:56   6  0  2  2   12  
5   168  0.916  2.24  4499:06   9  0  3  3    6  
6   209  0.903  2.84  4417:59   4  0  3  3    2  
7   130  0.933  1.96  3976:33   9  0  1  1    4  
8   156  0.924  2.36  3965:54   6  0  2  2    0  
9   170  0.922  2.43  4194:08   3  0  0  0    6  
10  168   0.92  2.17  4635:03   4  0  4  4    6  
11  167   0.92  2.62  3825:11   8  0  2  2   14  
12  137  0.918  2.04  4034:37  12  0  0  0   16  
13  138     --  1.97  4200:00  11  0  0  0    2  
14  133     --  1.90  4200:00  12  0  0  0    2  
15  166  0.923  2.39  4168:52   5  0  5  5    0  
16  130  0.917  1.89  4128:24  10  0  3  3   10  
17  187  0.911  2.57  4364:35   5  0  3  3    4  
18  161   0.91  2.24  4311:54   6  1  4  5   16  
19  170   0.91  2.47  4126:41   4  0  3  3   34  

[20 rows x 21 columns]

How many rows does it have?

In [14]:
goalieDf.shape

(4086, 21)

We have 4086 rows and 21 columns.  This makes sense because this dataframe contains every season stat for every goalie who has played in the NHL since 1950! While this dataframe contains lots of great information, we will want to add the age information to the data frame, as this is something we want to analyze.  The remainder of this notebook will be devoted to that process.

### Adding age information to the Goalie Statistics DataFrame:

There are two ways to go about this.  We could use the <a href="https://gitlab.com/dword4/nhlapi/-/blob/master/stats-api.md#people"> NHL's undocumented public API </a> in order to find the current age information, then convert it to age data for each season by subtracting a number from the current age depending on the year.  This may be doable, but would necessitate finding the "people ID" number for every single player in the dataframe, which would in turn require querying the teams to find the team roster ID's for each year, then using those rosters to extract the player ID's, then using the player ID's to get the bio information. The other possibility is to download the goalie biographical information (which contains date of birth) via excel exports from NHL's statistics webpage and then transforming the data and adding the current age for each player for each season.  We are going to do the latter.

In [16]:
bioBaseURL = "/Users/codypewarchuk/Downloads/Bio Info "
listOfBioDFs = []
for i in range(8):
    df = pd.read_excel(bioBaseURL + '(' + str(i) + ').xlsx')
    listOfBioDFs.append(df)

biosDf = pd.concat(listOfBioDFs, ignore_index = True)

Let's check out the first five rows in the bios dataframe.

In [17]:
biosDf.head()

Player Team S/C         DOB    Birth City S/P Ctry Ntnlty  Ht  \
0          Rob Zepp   --   L  1981-09-07   Scarborough  ON  CAN    CAN  74   
1      Jeff Zatkoff   --   L  1987-06-09       Detroit  MI  USA    USA  74   
2    Michael Zanier   --   L  1962-08-22         Trail  BC  CAN    CAN  71   
3  Artyom Zagidulin   --   L  1995-08-08  Magnitogorsk  --  RUS    RUS  74   
4         Matt Zaba   --   L  1983-07-14       Yorkton  SK  CAN    CAN  73   

    Wt  ... Round Overall 1st Season  HOF  GP   W   L   T  OT SO  
0  198  ...     4     110   20142015    N  10   5   2  --   0  0  
1  186  ...     3      74   20132014    N  48  18  21  --   4  1  
2  189  ...    --      --   19841985    N   3   1   1   1  --  0  
3  180  ...    --      --   20202021    N   1   0   0  --   0  0  
4  190  ...     8     231   20092010    N   1   0   0  --   0  0  

[5 rows x 21 columns]

In [18]:
biosDf.columns

Index(['Player', 'Team', 'S/C', 'DOB', 'Birth City', 'S/P', 'Ctry', 'Ntnlty',
       'Ht', 'Wt', 'Draft Yr', 'Round', 'Overall', '1st Season', 'HOF', 'GP',
       'W', 'L', 'T', 'OT', 'SO'],
      dtype='object')

The rows we care about for data analysis purposes are 'Player' (to integrate the tables together), 'DOB' (to determine the age), 'S/P' and 'Ntnlty' to determine where the goalies come from, 'Round' (to see how goalies drafted in different rounds have done), and 'HOF' (what makes a hall of famer!).

In [19]:
biosDf = biosDf[['Player', 'DOB', 'S/P', 'Ntnlty', 'Round', 'HOF']]
biosDf.head()

Player         DOB S/P Ntnlty Round HOF
0          Rob Zepp  1981-09-07  ON    CAN     4   N
1      Jeff Zatkoff  1987-06-09  MI    USA     3   N
2    Michael Zanier  1962-08-22  BC    CAN    --   N
3  Artyom Zagidulin  1995-08-08  --    RUS    --   N
4         Matt Zaba  1983-07-14  SK    CAN     8   N

Let's check the data type for each column.  Let's drop the day and month information for each player, because the age estimates do not need to be precise.  We will convert the column name to 'Birth Year' as well.

In [24]:
pd.to_datetime(biosDf['DOB'])
biosDf['DOB'] = pd.DatetimeIndex(biosDf['DOB']).year
biosDf.head()

Player   DOB S/P Ntnlty Round HOF
0          Rob Zepp  1981  ON    CAN     4   N
1      Jeff Zatkoff  1987  MI    USA     3   N
2    Michael Zanier  1962  BC    CAN    --   N
3  Artyom Zagidulin  1995  --    RUS    --   N
4         Matt Zaba  1983  SK    CAN     8   N

In [30]:
biosDf.rename(columns = {'DOB': 'Birth Year'}, inplace = True)
biosDf.head()

Player  Birth Year S/P Ntnlty Round HOF
0          Rob Zepp        1981  ON    CAN     4   N
1      Jeff Zatkoff        1987  MI    USA     3   N
2    Michael Zanier        1962  BC    CAN    --   N
3  Artyom Zagidulin        1995  --    RUS    --   N
4         Matt Zaba        1983  SK    CAN     8   N

Now all that's left to do is combining the dataframes.  We need to match up the player names, and add the column information.

In [73]:
birthYearList = []
stateList = []
nationList = []
rndList = []
hofList = []

for row in goalieDf['Player']:
    for i, bioRow in biosDf.iterrows():
        if row == bioRow[0]:
            birthYearList.append(bioRow[1])
            stateList.append(bioRow[2])
            nationList.append(bioRow[3])
            rndList.append(bioRow[4])
            hofList.append(bioRow[5])
            

goalieDf['Birth Year'] = birthYearList
goalieDf['S/P'] = stateList
goalieDf['Nationality'] = nationList
goalieDf['Draft Round'] = rndList
goalieDf['HOF'] = hofList

goalieDf.head()      

Player    Season Team S/C  GP  GS   W   L   T  OT  ...  SO  G  A  \
0   Braden Holtby  20152016  WSH   L  66  66  48   9  --   7  ...   3  0  1   
1  Martin Brodeur  20062007  NJD   L  78  78  48  23  --   7  ...  12  0  1   
2   Bernie Parent  19731974  PHI   L  73  73  47  13  12  --  ...  12  0  3   
3  Roberto Luongo  20062007  VAN   L  76  75  47  22  --   6  ...   5  0  2   
4  Evgeni Nabokov  20072008  SJS   L  77  77  46  21  --   8  ...   6  0  2   

   P  PIM Birth Year  S/P  Nationality  Draft Round  HOF  
0  1    6       1989   SK          CAN            4    N  
1  1   12       1972   QC          CAN            1    Y  
2  3   24       1945   QC          CAN           --    Y  
3  2   10       1979   QC          CAN            1    N  
4  2   12       1975   --          RUS            9    N  

[5 rows x 26 columns]

Now all that's left to do is use the birth year to make an age column.

In [79]:
goalieAges = []
for index, row in goalieDf.iterrows():
    temp = str(row[1])
    firstYearOfSeason = int(temp[:4])
    #row[21] is the 'Birth Year' for that row
    ageThatYear = firstYearOfSeason - row[21]
    goalieAges.append(ageThatYear)

goalieDf['Age'] = goalieAges

In [93]:
goalieDf.sort_values(by = ['Player','Season'], axis = 0, inplace=True)

In [94]:
goalieDf.head(20)

Player    Season Team S/C  GP  GS   W   L   T  OT  ...  G  A  P  \
1768     Aaron Dell  20162017  SJS   L  20  17  11   6  --   1  ...  0  0  0   
1390     Aaron Dell  20172018  SJS   L  29  22  15   5  --   4  ...  0  1  1   
1949     Aaron Dell  20182019  SJS   L  25  20  10   8  --   4  ...  0  0  0   
1698     Aaron Dell  20192020  SJS   L  33  30  12  15  --   3  ...  0  0  0   
3424     Aaron Dell  20202021  NJD   L   7   5   1   5  --   0  ...  0  0  0   
3085  Adam Berkhoel  20052006  ATL   L   9   7   2   4  --   1  ...  0  0  0   
4023    Adam Hauser  20052006  LAK   L   1   0   0   0  --   0  ...  0  0  0   
3359     Adam Munro  20032004  CHI   L   7   7   1   5   1  --  ...  0  0  0   
2870     Adam Munro  20052006  CHI   L  10  10   3   5  --   2  ...  0  1  1   
3252    Adam Werner  20192020  COL   L   2   1   1   1  --   0  ...  0  0  0   
3553    Adam Wilcox  20172018  BUF   L   1   0   0   1  --   0  ...  0  0  0   
3323      Adin Hill  20172018  ARI   L   4   4   1   3  --   0  ...  0  0  0   
2304      Adin Hill  20182019  ARI   L  13  11   7   5  --   0  ...  0  0  0   
3000      Adin Hill  20192020  ARI   L  13   9   2   4  --   3  ...  0  0  0   
2008      Adin Hill  20202021  ARI   L  19  17   9   9  --   1  ...  0  0  0   
4007      Al Jensen  19801981  DET   L   1   1   0   1   0  --  ...  0  0  0   
2211      Al Jensen  19811982  WSH   L  26  19   8   8   4  --  ...  0  2  2   
832       Al Jensen  19821983  WSH   L  40  39  22  12   6  --  ...  0  0  0   
625       Al Jensen  19831984  WSH   L  43  42  25  13   3  --  ...  0  0  0   
1951      Al Jensen  19841985  WSH   L  14  13  10   3   1  --  ...  0  0  0   

     PIM  Birth Year S/P  Nationality  Draft Round  HOF  Age  
1768   0        1989  AB          CAN           --    N   27  
1390   0        1989  AB          CAN           --    N   28  
1949   0        1989  AB          CAN           --    N   29  
1698   0        1989  AB          CAN           --    N   30  
3424   0        1989  AB          CAN           --    N   31  
3085   0        1981  MN          USA            8    N   24  
4023   0        1980  MN          USA            3    N   25  
3359   2        1982  ON          CAN            1    N   21  
2870   0        1982  ON          CAN            1    N   23  
3252   0        1997  --          SWE            5    N   22  
3553   0        1992  MN          USA            6    N   25  
3323   0        1996  BC          CAN            3    N   21  
2304   0        1996  BC          CAN            3    N   22  
3000   2        1996  BC          CAN            3    N   23  
2008   0        1996  BC          CAN            3    N   24  
4007   0        1958  ON          CAN            2    N   22  
2211   6        1958  ON          CAN            2    N   23  
832    6        1958  ON          CAN            2    N   24  
625   22        1958  ON          CAN            2    N   25  
1951   6        1958  ON          CAN            2    N   26  

[20 rows x 27 columns]

Okay great!  Let's export the dataframe to a .csv so we can start analyzing it!

In [95]:
goalieDf.to_csv('GoalieStats1950-2020.csv')